src/mpid/ch4/netmod/include/netmod_am_fallback_send.h

```diff
@@ -27,6 +29,16 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_mpi_isend(const void *buf,
     MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(vci_src).lock);
     mpi_errno = MPIDIG_mpi_isend(buf, count, datatype, rank, tag, comm, context_offset, addr,
                                  vci_src, vci_dst, request, syncflag, errflag);
+    /* if the parent_cc_ptr exists */
+    if (parent_cc_ptr) {
+        if (MPIR_Request_is_complete(*request)) {
+            /* if the request is already completed, decrement the parent counter */
+            MPIR_cc_dec(parent_cc_ptr);
+        } else {
+            /* if the request is not done yet, assign the completion pointer to the parent one and it will be decremented later */
+            (*request)->dev.completion_notification = parent_cc_ptr;
+        }
+    }
```

Comment:
can this be done by the caller instead? it might be cleaner, assuming we have access to the parent request when each operation is issued. something like:
```c
MPIDI_NM_mpi_isend(..., &childreq);
if (childreq && !MPIR_Request_is_complete(childreq)) {
    childreq->dev.completion_notification = parentreq->cc_ptr;
} else {
    MPID_Request_complete(parentreq);
}
```

Reply:
If I understand your suggestion correctly, we would exit the lock around the child request creation which might not work when async progress is enabled (as you suggested a few months ago :-) )
The reason is that if you are not in the same lock section, you have no guarantee the completion notification will be triggered: you might complete the request between the release of the lock at creation and the assignment of `childreq->dev.completion_notification = parentreq->cc_ptr;`